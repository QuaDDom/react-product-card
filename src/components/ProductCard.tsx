import React from 'react'
import { CSSProperties } from 'react';
import styles from '../styles/styles.module.css';
import ProductTitle from './ProductTitle';
import ProductImage from './ProductImage';
import ProductButtons from './ProductButtons';
import noImage from '../assets/no-image.jpg';
import { ProductContext } from '../context/ProductContext';
import { useProduct } from '../hooks/useProduct';
import { InitialValuesI, OnChangeArgsI, ProductCardHandlersI } from '../interfaces/interfaces';


interface PropsI {
    // children?: ReactElement | ReactElement[],
    children: ( args: ProductCardHandlersI )=> JSX.Element,
    product?: ProductI,
    className?: string,
    style?: CSSProperties,
    onChange?: (args: OnChangeArgsI)=> void,
    value?: number,
    initialValues?: InitialValuesI
}

interface ProductI{
    id: string,
    title: string,
    image?: string,
    count?: number
}

const productDefault = {
    id: '1',
    title: 'No Title',
    image: noImage
}

export default function ProductCard({ children, product = productDefault, className, style, value, initialValues }:PropsI) {
    const { Provider } = ProductContext;
    const {counter, handleClick, maxCount, isMaxCountReached, reset} = useProduct({
        product,
        value,
        initialValues
    });

    return ( 
        <Provider value={{
            counter,
            handleClick,
            maxCount,
            product
        }}>
            <div  className={`${styles.productCard} ${className}`} style={style || {}}>
                { 
                    children({
                        count: counter,
                        isMaxCountReached,
                        maxCount,
                        product,
                        handleClick,
                        reset
                    }) 
                }
            </div>
        </Provider>
    )
}

ProductCard.Title = ProductTitle;
ProductCard.Image = ProductImage;
ProductCard.Buttons = ProductButtons;