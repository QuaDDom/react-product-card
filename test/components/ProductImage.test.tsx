import React from 'react'
import renderer from 'react-test-renderer'
import ProductCard, { ProductImage } from '../../src/components'
import { product2 } from '../data/products';

describe('ProductImage', ()=>{
    test('show image without ProductCard component', ()=>{
        const wrapper = renderer.create(
            <ProductImage img="test.png"/>
        )
        expect( wrapper.toJSON ).toMatchSnapshot();
    });
    test('Show image in ProductCard component', ()=>{
        const wrapper = renderer.create(
            <ProductCard product={product2}>
                {
                    ()=>(
                        <ProductImage/>
                    )
                }
            </ProductCard>
        )
        let tree = wrapper.toJSON();
        console.log(tree)
    })
})