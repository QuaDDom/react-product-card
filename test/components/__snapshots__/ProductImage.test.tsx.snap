// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage show image without ProductCard component 1`] = `[Function]`;
